dev-client/src/screens/AppScaffold.tsx

```diff
@@ -1,80 +1,137 @@
+};
 
-    if (config.hideBack) {
-      options.headerBackVisible = false;
-    }
+const DefaultHeader = ({
+  name,
+  ...props
+}: React.ComponentProps<typeof NavigationHeaderTitle> & {name: string}) => {
+  const {t} = useTranslation();
+  return (
+    <NavigationHeaderTitle {...props}>
+      {t('screens.' + name) ?? name}
+    </NavigationHeaderTitle>
+  );
+};
 
-    options.headerLeft = config.leftIcon;
-    options.headerRight = config.rightIcon;
-
-    return (
-      <Stack.Screen
-        name={name}
-        component={config.screen}
-        initialParams={initialParams}
-        key={name}
-        options={({route}) => {
-          if (config.paramTitle !== undefined) {
-            options.headerTitle = config.paramTitle({t, route});
-          } else {
-            options.headerTitle = t('screens.' + name) ?? undefined;
-          }
-          return options;
-        }}
-      />
-    );
+const defineScreen = <
+  Name extends ScreenName,
+  Props extends RootStackParamList[Name],
+>(
+  name: Name,
+  {options, View}: ScreenDefinition<Props>,
+) => {
+  const wrappedOptions = (
+    props: RootStackScreenProps,
+  ): NativeStackNavigationOptions => {
+    const {HeaderTitle = undefined, ...origOptions} =
+      options === undefined ? {} : options(props.route.params as Props);
+    const CastTitle = HeaderTitle as HeaderTitleType<object> | undefined;
+    return {
+      ...origOptions,
+      headerTitle: ({tintColor}) =>
+        CastTitle !== undefined ? (
+          <CastTitle {...props.route.params} tintColor={tintColor} />
+        ) : (
+          <DefaultHeader name={name} tintColor={tintColor} />
+        ),
+    };
   };
-}
+
+  const Component = (props: RootStackScreenProps) => (
+    <View {...((props.route.params ?? {}) as Props & object)} />
+  );
+
+  return [
+    name,
+    <RootStack.Screen
+      name={name}
+      key={name}
+      component={Component}
+      options={wrappedOptions}
+    />,
+  ];
+};
+
+const definedScreens = Object.fromEntries(
+  Object.entries(screens).map(([name, screen]) =>
+    defineScreen(name as ScreenName, screen as ScreenDefinition),
+  ),
+);
+
+export const useNavigation = <Name extends ScreenName = ScreenName>() =>
+  useNavigationNative<NavigationProp<RootStackParamList, Name>>();
 
 export default function AppScaffold() {
-  const {t} = useTranslation();
   const dispatch = useDispatch();
   const hasToken = useSelector(state => state.account.hasToken);
   const currentUser = useSelector(state => state.account.currentUser.data);
   // using theme hook here because react-navigation doesn't take nativebase utility props
   const {colors} = useTheme();
 
-  function filterLogin([_name, config]: ScreenMapArgs) {
-    if (currentUser === null) {
-      return config.loggedOut === true;
-    } else {
-      return config.loggedOut !== true;
-    }
-  }
+  const filterLogin = (name: ScreenName) =>
+    (currentUser === null) === (name === 'LOGIN');
```

Comment:
nit: Do you think it makes sense to keep more than just one "logged out" screen? I think we will likely have several screens where the user doesn't have to log in - but maybe we can just deal with that as it arises.

Reply:
i'm not sure what this will look like eventually, because the actual intended workflow is that the whole app should be usable without logging in right?

Reply:
(minus collaborative features)

Reply:
Yeah, I was just thinking there will always be a bunch of screens where you have to login for them to make sense. But let's not worry about that now.